docs/contribute/testing.adoc

```diff
@@ -0,0 +1,423 @@
+$ oscap xccdf eval *--rule X* --profile Y \
+    /usr/share/xml/scap/ssg/content/ssg-fedora-ds.xml
+----
+
+
+=== Deciding where to put a new test
+In our example, we are changing the `xccdf` module, therefore we will look for
+its subdirectory in the link:../../tests[tests] direcotory and we will find it
+at the following link: link:../../tests/API/XCCDF[tests/API/XCCDF]. There are
+multiple other subdirectories with tests and we need to decide where to put our
+test for the evaluation of a single rule. From all these subdirectories the
+most suitable one appears to be the
+link:../../tests/API/XCCDF/unittests[tests/API/XCCDF/unittests] directory as
+all the general XCCDF tests for `oscap xccdf eval` are located there. We will
+add our new test into this subdirectory.
+
+
+==== Scenario A: There is a suitable directory for my new test
+This will happen most of the times. As stated above in our example we will place
+our test into the
+link:../../tests/API/XCCDF/unittests[tests/API/XCCDF/unittests] directory.
+
+
+==== Scenario B: There is no suitable directory for my new test
+This might happen if your test covers a part of the OpenSCAP which has no tests
+at all. In this case you would need to add a new directory with suitable name
+into the link:../../tests[tests] directory structure and create/update
+Makefiles.
+
+To have an example also for this scenario, let's suppose we want to add the
+`foo` subdirectory into the link:../../tests/API/XCCDF[tests/API/XCCDF]
+directory. We need to:
+
+. Create the `foo` subdirectory in the
+   link:../../tests/API/XCCDF[tests/API/XCCDF] directory.
+. Add a new entry for `foo` subdirectory into the `SUBDIRS` variable in
+   the link:../../tests/API/XCCDF/Makefile.am[tests/API/XCCDF/Makefile.am].
+   When adding a new entry please keep in mind that all entries are sorted in
+   the alphabetical order.
+. Create `Makefile.am` inside the `tests/API/XCCDF/foo` directory with the
+   following contents:
+[[app-listing]]
+[source,bash]
+----
+DISTCLEANFILES = *.log *.out* oscap_debug.log.*
+CLEANFILES = *.log *.out* oscap_debug.log.*
+
+TESTS_ENVIRONMENT = \
+	builddir=$(top_builddir) \
+	OSCAP_FULL_VALIDATION=1 \
+	$(top_builddir)/run
+
+# This is the entry script which should be responsible for running
+# all the tests from the 'foo' directory.
+TESTS = all.sh
+
+# This variable must contain all the files which are required to run tests.
+EXTRA_DIST = \
+	all.sh \
+	my_test_script.sh \
+	my_data.xml
+----
+
+[start=4]
+. Create an entry script, for example `all.sh`, inside the `tests/API/XCCDF/foo`
+   directory. This script should be responsible for executing all your test
+   scripts with specified arguments, for example:
+[[app-listing]]
+[source,bash]
+----
+#!/usr/bin/env bash
+
+set -o pipefail
+
+. ../../../test_common.sh
+
+function test_foo() {
+	...
+	return 0
+}
+
+test_init test_api_xccdf_foo.log
+
+test_run "Description 1" $srcdir/my_test_script.sh my_data.xml
+test_run "Description 2" test_foo
```

Comment:
again, we are trying to do just one test_run per file. and I will say that I don't like the testing harness we have, it's ridiculous and just makes things harder... we should rely on cmake/autotools to collect the right outputs, not write bash magic to fix it...